core/src/cluster_info.rs

```diff
@@ -1102,22 +1103,59 @@ impl ClusterInfo {
         self.push_message(CrdsValue::new_signed(message, &self.keypair));
     }
 
-    pub fn push_vote(&self, tower_index: usize, vote: Transaction) {
+    pub fn push_vote(&self, tower: &[Slot], vote: Transaction) {
+        debug_assert!(tower.iter().tuple_windows().all(|(a, b)| a < b));
         let now = timestamp();
-        let vote = Vote::new(&self.id(), vote, now);
-        let vote_ix = {
-            let r_gossip =
+        // Find a crds vote which is evicted from the tower, and recycle its
+        // vote-index. This can be either an old vote which is popped off the
+        // deque, or recent vote which has expired before getting enough
+        // confirmations.
+        // If all votes are still in the tower, add a new vote-index. If more
+        // than one vote is evicted, the one with most recent wallclock is
+        // returned because that had got fewer confirmations.
+        // TODO: When there are more than one vote evicted from the tower, only
+        // one crds vote is overwritten here. Decide what to do with the rest.
+        let mut num_crds_votes = 0;
+        let self_pubkey = self.id();
+        // Returns true if the tower does not contain the vote.slot.
+        let should_evict_vote = |vote: &Vote| -> bool {
+            match vote.slot() {
+                Some(slot) => !tower.contains(&slot),
+                None => {
+                    error!("crds vote with no slots!");
+                    true
+                }
+            }
+        };
+        let vote_index = {
+            let gossip =
                 self.time_gossip_read_lock("gossip_read_push_vote", &self.stats.push_vote_read);
-            let current_votes: Vec<_> = (0..crds_value::MAX_VOTES)
-                .filter_map(|ix| r_gossip.crds.lookup(&CrdsValueLabel::Vote(ix, self.id())))
-                .collect();
-            CrdsValue::compute_vote_index(tower_index, current_votes)
+            (0..MAX_LOCKOUT_HISTORY as u8)
+                .filter_map(|ix| {
+                    let vote = CrdsValueLabel::Vote(ix, self_pubkey);
+                    let vote = gossip.crds.lookup(&vote)?;
+                    num_crds_votes += 1;
+                    match &vote.data {
+                        CrdsData::Vote(_, vote) if should_evict_vote(vote) => {
+                            Some((vote.wallclock, ix))
+                        }
+                        CrdsData::Vote(_, _) => None,
+                        _ => panic!("this should not happen!"),
+                    }
+                })
+                .max() // Take the evicted vote with the most recent wallclock.
```

Comment:
hmmm thinking more on this, I'm a bit worried about evicting the most recent wallclock because of the implications for optimistic confirmation which relies on giving those votes enough time to propagate through the rest of the network.  Please see: https://github.com/solana-labs/solana/blob/master/core/src/cluster_info_vote_listener.rs#L578-L596

So where before, if we had two forks

```
1, 2, 3, 4, 5, 6
     |
     20
```
Let's say the order of events is:

1)  the validator first voted on slots 1, 2, 3, 4 ,5, 6 on the first fork
2) The lock on slot 3 is `16`, so they can't vote on fork 2 until slot 20
3) The validator now votes on slot 20

In the original code, slot 20 would pop off the vote for slot 3 (which already has been sitting in gossip for 17 slots), giving the later votes 4, 5, 6 to propagate. In the new code, this vote for 20 would immediately pop off slot `6`, which hasn't had as long to propagate.

Reply:
> In the new code, this vote for 20 would immediately pop off slot 6, which hasn't had as long to propagate.

So, one thing I do not understand is if a validator has switched to a different fork, why should we want to propagate a recently expired vote (with no or only few confirmations) from an older fork which this validator has already switched off?
Shouldn't we instead favor propagating votes with more confirmations?

> In this respect, I think it may be safer to stay as close to the original code to place the vote based on its position in the tower.

The current code is actually giving less time to votes to propagate. It keeps 29 oldest votes by wallclock, and then
only overrides the 30st one each time. (i.e a tally of only two most recent votes):
https://github.com/solana-labs/solana/blob/85fdbd87b/core/src/crds_value.rs#L612-L617

So once it hits `MAX_VOTES`, a vote is kicked out as soon as there are only 2 more recent votes.

Reply:
>The current code is actually giving less time to votes to propagate. It keeps 29 oldest votes by wallclock, and then
only overrides the 30st one each time. (i.e a tally of only two most recent votes):
https://github.com/solana-labs/solana/blob/85fdbd87b/core/src/crds_value.rs#L612-L617

Right that's assuming we keep pushing to the stack in such a way that the stack is growing right? i.e. the tower index is always `30` on each update. But in the above example, let's say the the tower index on the vote `6` is `N`. Then after vote `20`, because `3, 4, 5, 6` are all removed from the tower, the tower index for the vote `20` update is now `N - 3`, replacing vote `3` in the stack. In the new code we would remove slot `6`.

>So, one thing I do not understand is if a validator has switched to a different fork, why should we want to propagate a recently expired vote (with no or only few confirmations) from an older fork which this validator has already switched off?
Shouldn't we instead favor propagating votes with more confirmations?

Yeah, so the consideration here is for the optimistic confirmation code here: https://github.com/solana-labs/solana/blob/master/core/src/cluster_info_vote_listener.rs#L578-L596, which only counts a vote toward a slot `S`'s optimistic confirmation if the *last* slot in the vote ==  `S`. Thus the reason to keep around an expired vote is to give the last vote maximal time to propagate, just in case the previous fork was the major fork. 

So in the example above, slots `3, 4, 5, 6` were all removed, so we're just deciding if it's better to remove `3` or `6`. My feeling is because the vote for `6` potentially carries more information (it includes earlier missing votes between 3 and 6 that may not have landed, and is for a more specific fork than just `3`).  It may not make a big difference in the general case, but it differs from the behavior today. Maybe @aeyakovenko can comment if there are additional subtlties here.

Reply:
updated the code to boot the oldest vote by wallclock